neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsList.java

```diff
@@ -0,0 +1,157 @@
+     */
+    @Nonnull
+    private final PersistentEObject owner;
+
+    /**
+     * Constructs a new {@code ContentsList}.
+     *
+     * @param owner    the owner of this list
+     * @param features the containment features that are handled by this list
+     */
+    protected ContentsList(PersistentEObject owner, EStructuralFeature[] features) {
+        super(owner, features);
+        this.owner = owner;
+    }
+
+    /**
+     * Returns an empty {@code ContentsList}.
+     *
+     * @param <E> the type of elements in this list
+     *
+     * @return an empty list
+     */
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public static <E extends EObject> ContentsList<E> empty() {
+        return (ContentsList<E>) EMPTY;
+    }
+
+    /**
+     * Creates a new {@code ContentsList} with the given {@code owner}.
+     *
+     * @param owner the owner of this list
+     * @param <E>   the type of elements in this list
+     *
+     * @return a new list
+     */
+    @Nonnull
+    public static <E extends EObject> ContentsList<E> newList(PersistentEObject owner) {
+        List<EStructuralFeature> features = owner.eClass().getEAllStructuralFeatures();
+        EStructuralFeature[] containments = EClassImpl.FeatureSubsetSupplier.class.cast(features).containments();
+
+        return nonNull(containments)
+                ? new ContentsList<>(owner, containments)
+                : ContentsList.empty();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public E get(int index) {
+        checkNotNull(eStructuralFeatures, "index=" + index + ", size=0");
+
+        // Find the feature to look for
+        int featureSize = 0;
+
+        for (EStructuralFeature feature : eStructuralFeatures) {
+            int localFeatureSize = feature.isMany()
+                    ? owner.eStore().size(owner, feature)
+                    : owner.eStore().isSet(owner, feature) ? 1 : 0;
+
+            featureSize += localFeatureSize;
+
+            if (featureSize > index) {
+                // The correct feature has been found
+                int featureIndex = index - (featureSize - localFeatureSize);
+                return (E) owner.eStore().get(owner, feature, featureIndex);
+            }
+        }
+
+        throw new IndexOutOfBoundsException("index=" + index + ", size=" + featureSize);
+    }
+
+    @Override
+    protected ListIterator<E> newResolvingListIterator() {
+        return new ContentsListIteratorResolver<>(owner, eStructuralFeatures);
+    }
+
+    @Override
+    protected ListIterator<E> newNonResolvingListIterator() {
+        return new ContentsListIterator<>(owner, eStructuralFeatures);
+    }
+
+    @Override
+    protected final boolean useIsSet() {
+        return false;
+    }
+
+    /**
+     * An empty {@code ContentsList}.
+     *
+     * @param <E> the type of elements in this list
+     */
+    @ParametersAreNullableByDefault
+    private static class Empty<E extends EObject> extends ContentsList<E> {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This class has 7 parents which is greater than 5 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)